pico/services/src/main/java/io/helidon/pico/services/DefaultPicoServices.java

```diff
@@ -0,0 +1,528 @@
+                initializeServices();
+            } catch (Throwable t) {
+                state.lastError(t);
+                initializingServicesStarted.set(false);
+                if (t instanceof PicoException) {
+                    throw (PicoException) t;
+                } else {
+                    throw new PicoException("failed to initialize: " + t.getMessage(), t);
+                }
+            } finally {
+                state.finished(true);
+                initializingServicesFinished.set(true);
+            }
+        }
+
+        DefaultServices thisServices = services.get();
+        if (thisServices == null) {
+            throw new PicoException("must reset() after shutdown()");
+        }
+        return Optional.of(thisServices);
+    }
+
+    @Override
+    public Optional<Map<String, ActivationResult>> shutdown() {
+        Map<String, ActivationResult> result = Map.of();
+        DefaultServices current = services.get();
+        if (services.compareAndSet(current, null) && current != null) {
+            State currentState = state.clone().currentPhase(Phase.PRE_DESTROYING);
+            log("started shutdown");
+            result = doShutdown(current, currentState);
+            log("finished shutdown");
+        }
+        return Optional.ofNullable(result);
+    }
+
+    private Map<String, ActivationResult> doShutdown(DefaultServices services,
+                                                     State state) {
+        long start = System.currentTimeMillis();
+
+        ThreadFactory threadFactory = r -> {
+            Thread thread = new Thread(r);
+            thread.setDaemon(false);
+            thread.setPriority(Thread.MAX_PRIORITY);
+            thread.setName(PicoServicesConfig.NAME + "-shutdown-" + System.currentTimeMillis());
+            return thread;
+        };
+
+        Shutdown shutdown = new Shutdown(services, state);
+        ExecutorService es = Executors.newSingleThreadExecutor(threadFactory);
+        long finish;
+        try {
+            return es.submit(shutdown)
+                    // https://github.com/helidon-io/helidon/issues/6434: have an appropriate timeout config for this
+//                    .get(cfg.activationDeadlockDetectionTimeoutMillis(), TimeUnit.MILLISECONDS);
+                    .get();
+        } catch (Throwable t) {
+            finish = System.currentTimeMillis();
+            errorLog("error during shutdown (elapsed = " + (finish - start) + " ms)", t);
+            throw new PicoException("error during shutdown", t);
+        } finally {
+            es.shutdown();
+            state.finished(true);
+            finish = System.currentTimeMillis();
+            log("finished shutdown (elapsed = " + (finish - start) + " ms)");
+        }
+    }
+
+    /**
+     * Will attempt to shut down in reverse order of activation, but only if activation logs are retained.
+     */
+    private class Shutdown implements Callable<Map<String, ActivationResult>> {
```

Comment:
wrong order of types/methods in this class.
also non-static inner classes are a bit confusing, as it is hard to read what is used from the containing instance, and what is handled over parameters to constructor.